hera_cal/lst_stack/calibration.py

```diff
@@ -0,0 +1,371 @@
+
+
+    Parameters:
+    ----------
+        stack : LSTStack
+            The LSTStack object to calibrate
+        model : np.ndarray
+            The reference model to calibrate the data to. The model should have the same
+            number of baselines, frequencies, and polarizations as the data in stack.
+        all_reds : list of list of tuples
+            A list of lists of redundant baseline groups. Each element of the list is a list of tuples
+            containing the redundant baseline groups. Each tuple contains the antenna numbers
+            of the antennas in the redundant baseline group.
+        inpaint_bands : tuple of slices, default=(slice(0, None, None),)
+            Defines the frequency bands to use for smoothing. Each slice object in the tuple
+            defines a separate frequency band to use for smoothing. The default is to smooth over
+            all frequencies at once.
+        auto_stack : LSTStack, default=None
+            An LSTStack object containing the stack of auto-correlations. If provided, the
+            auto-correlations will be calibrated using the same gains calculated from the
+            cross-correlations. Note that the auto-correlations are not used to calculate the
+            gains if provided.
+        run_amplitude_cal : bool, default=True
+            Boolean flag to run amplitude calibration.
+        run_phase_cal : bool, default=True
+            Boolean flag to run tip-tilt calibration.
+        smoothing_scale : float, default=10e6
+            The scale of the smoothing function used to smooth the gains. This is the width of the
+            smoothing function in Hz.
+        eigenval_cutoff : float, default=1e-12
+            The cutoff for the eigenvalues of the DPSS eigenvectors.
+        calibrate_inplace : bool, default=True
+            Boolean flag to calibrate the data in place.
+        return_gains : bool, default=True
+            Boolean flag to return the gains. If set to False, the function will return an empty
+            dictionary for the gains.
+
+    Returns:
+    -------
+        calibration_parameters : dict
+            A dictionary containing the calibration parameters. The keys are as follows:
+                - 'A_J{pol}' : The amplitude calibration parameters for each polarization.
+                - 'T_J{pol}' : The tip-tilt calibration parameters for each polarization
+        gains : dict
+            A dictionary containing the calibrated gains for each baseline. The keys are tuples
+            containing the antenna numbers and polarization of the baseline. If return_gains is
+            set to False, this dictionary will be empty.
+    """
+    # Assert some calibration done
+    assert (
+        run_amplitude_cal or run_phase_cal
+    ), "At least one calibration mode must be used"
+    assert (
+        stack.data.shape[1:] == model.shape
+    ), "Model must have the same number of antpairs/freqs/pols as stack.data"
```

Comment:
Yeah, it is better practice. It's not that it will really make a difference here, but `assert` is meant to be used where you expect something to _never fail_ (not dependent on user error etc.). Assertions are possible to _ignore_ when compiling python with a certain flag for optimization, so they're not recommended for catching when users pass bad parameters etc.

Reply:
Changed these to ValueErrors